File: steps/7/test/java/eightyDays/java/AccountingTest.java
package eightyDays.java;

import eightyDays.java.account.Current;
import eightyDays.java.account.Saving;
import org.junit.Test;

import java.math.BigDecimal;
import java.util.UUID;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

public class AccountingTest {

    @Test
    public void testSavingAccount() {
        Saving saving = new Saving(BigDecimal.valueOf(500), UUID.randomUUID());
        assertEquals(BigDecimal.ZERO, saving.getBalance());
        assertEquals(BigDecimal.valueOf(400),saving.deposit(BigDecimal.valueOf(400)).getBalance());
        assertEquals(BigDecimal.valueOf(101), saving.deposit(BigDecimal.valueOf(400)).withdraw(BigDecimal.valueOf(299)).getBalance());
    }

    @Test
    public void testSavingOverdraw() {
        Saving saving = new Saving(BigDecimal.valueOf(500), UUID.randomUUID());
        assertEquals(BigDecimal.ZERO, saving.getBalance());
        assertEquals(BigDecimal.valueOf(400),saving.deposit(BigDecimal.valueOf(400)).getBalance());
        assertEquals(BigDecimal.valueOf(100), saving.deposit(BigDecimal.valueOf(400)).withdraw(BigDecimal.valueOf(300)).getBalance());
        try {
            saving.deposit(BigDecimal.valueOf(400)).withdraw(BigDecimal.valueOf(401));
            fail("no exception thrown");
        } catch (RuntimeException exception) {
            assertEquals("Withdraw not possible due to insufficient getBalance", exception.getMessage());
        }
    }

    @Test
    public void testSavingOverWithdrawLimit() {
        Saving saving = new Saving(BigDecimal.valueOf(200), UUID.randomUUID());
        assertEquals(BigDecimal.ZERO, saving.getBalance());
        assertEquals(BigDecimal.valueOf(100), saving.deposit(BigDecimal.valueOf(300)).withdraw(BigDecimal.valueOf(200)).getBalance());
        try {
            saving.deposit(BigDecimal.valueOf(400)).withdraw(BigDecimal.valueOf(201));
            fail("no exception thrown");
        } catch (RuntimeException exception) {
            assertEquals("Value over withdraw limit", exception.getMessage());
        }
    }

    @Test
    public void testCurrent() {
        Current current = new Current(BigDecimal.valueOf(100), BigDecimal.valueOf(1.5), BigDecimal.valueOf(40), UUID.randomUUID());
        assertEquals(BigDecimal.ZERO, current.getBalance());
        assertEquals(BigDecimal.valueOf(10.5), current.deposit(BigDecimal.valueOf(12)).getBalance());
        assertEquals(BigDecimal.valueOf(7.0), current.deposit(BigDecimal.valueOf(12)).withdraw(BigDecimal.valueOf(2)).getBalance());
    }
    @Test
    public void testCurrentWithdrawOverLimit() {
        Current current = new Current(BigDecimal.valueOf(100), BigDecimal.valueOf(1.5), BigDecimal.valueOf(40), UUID.randomUUID());
        try {
            current.withdraw(BigDecimal.valueOf(400));
            fail("no exception thrown");
        } catch (RuntimeException exception) {
            assertEquals("Value over withdraw limit", exception.getMessage());
        }
    }
}

File: steps/7/test/scala/eightyDays/scala/AccountingTest.scala
package eightyDays.scala

;

import java.util.UUID

import eightyDays.scala.account.{Current, Saving}
import org.scalatest.WordSpec

class AccountingTest extends WordSpec {
  "A saving account" when {
    "deposit and withdraw" in {
      val testee = Saving(500)(UUID.randomUUID())
      assert(0 === testee.balance)
      assert(400 === testee.deposit(400).balance)
      assert(101 === testee.deposit(400).withdraw(299).balance)
    }
    "overdraw" in {
      val testee = Saving(500)(UUID.randomUUID())
      assert(0 === testee.balance)
      assert(400 === testee.deposit(400).balance)
      assert(100 === testee.deposit(400).withdraw(300).balance)
      assert("Withdraw not possible due to insufficient getBalance" === intercept[RuntimeException] {
        testee.deposit(400).withdraw(401)
      }.getMessage)
    }
    "over withdraw limit" in {
      val testee = Saving(200)(UUID.randomUUID())
      assert(0 === testee.balance)
      assert(100 === testee.deposit(300).withdraw(200).balance)
      assert("Value over withdraw limit" === intercept[RuntimeException] {
        testee.deposit(400).withdraw(201)
      }.getMessage)
    }
  }
  "A current account" when {
    "deposit and withdraw" in {
      val testee = Current(100)(1.5, 40)(UUID.randomUUID())
      assert(0 === testee.balance)
      assert(10.5 === testee.deposit(12).balance)
      assert(7.0 === testee.deposit(12).withdraw(2).balance)
    }
    "withdraw over limite" in {
      val testee = Current(100)(1.5, 40)(UUID.randomUUID())
      assert("Value over withdraw limit" === intercept[RuntimeException] {
        testee.withdraw(400)
      }.getMessage)
    }
  }
}
